r/src/datatype.cpp

```diff
@@ -20,86 +20,163 @@
 #if defined(ARROW_R_WITH_ARROW)
 #include <arrow/type.h>
 
-// [[arrow::export]]
-bool shared_ptr_is_null(SEXP xp) {
-  return reinterpret_cast<std::shared_ptr<void>*>(R_ExternalPtrAddr(xp))->get() ==
-         nullptr;
-}
+namespace cpp11 {
+template <>
+std::string r6_class_name<arrow::DataType>(const std::shared_ptr<arrow::DataType>& type) {
+  using arrow::Type;
+
+  switch (type->id()) {
+    case Type::NA:
+      return "Null";
+    case Type::BOOL:
+      return "Boolean";
+    case Type::UINT8:
+      return "UInt8";
+    case Type::UINT16:
+      return "UInt16";
+    case Type::UINT32:
+      return "UInt32";
+    case Type::UINT64:
+      return "UInt64";
+
+    case Type::INT8:
+      return "Int8";
+    case Type::INT16:
+      return "Int16";
+    case Type::INT32:
+      return "Int32";
+    case Type::INT64:
+      return "Int64";
+
+    case Type::HALF_FLOAT:
+      return "Float16";
+    case Type::FLOAT:
+      return "Float32";
+    case Type::DOUBLE:
+      return "Float64";
+
+    case Type::STRING:
+      return "Utf8";
+    case Type::LARGE_STRING:
+      return "LargeUtf8";
+
+    case Type::BINARY:
+      return "Binary";
+    case Type::FIXED_SIZE_BINARY:
+      return "FixedSizeBinary";
+    case Type::LARGE_BINARY:
+      return "LargeBinary";
+
+    case Type::DATE32:
+      return "Date32";
+    case Type::DATE64:
+      return "Date64";
+    case Type::TIMESTAMP:
+      return "Timestamp";
+
+    case Type::TIME32:
+      return "Time32";
+    case Type::TIME64:
+      return "Time64";
+
+    case Type::DECIMAL:
+      return "Decimal128Type";
+
+    case Type::LIST:
+      return "ListType";
+    case Type::LARGE_LIST:
+      return "LargeListType";
+    case Type::FIXED_SIZE_LIST:
+      return "FixedSizeListType";
+
+    case Type::STRUCT:
+      return "StructType";
+    case Type::DICTIONARY:
+      return "DictionaryType";
+
+    default:
+      break;
+  }
 
-// [[arrow::export]]
-bool unique_ptr_is_null(SEXP xp) {
-  return reinterpret_cast<std::unique_ptr<void>*>(R_ExternalPtrAddr(xp))->get() ==
-         nullptr;
+  return "DataType";
+
+  // switch(names(Type)[self$id + 1],
+  //
+  //        INTERVAL = stop("Type INTERVAL not implemented yet"),
+  //        SPARSE_UNION = stop("Type SPARSE_UNION not implemented yet"),
+  //        DENSE_UNION = stop("Type DENSE_UNION not implemented yet"),
+  //        MAP = stop("Type MAP not implemented yet"),
+  //        EXTENSION = stop("Type EXTENSION not implemented yet"),
+  //        DURATION = stop("Type DURATION not implemented yet"),
```

Comment:
IIUC, if an `IntervalType` is passed to R we will now get an opaque `DataType` instead of an error message. Whatever the case, please add a comment describing behavior if one of these type ids is encountered